crates/core-manager/src/core_range.rs

```diff
@@ -0,0 +1,257 @@
+use range_set_blaze::RangeSetBlaze;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::fmt::{Debug, Display, Formatter};
+use std::str::FromStr;
+use thiserror::Error;
+
+#[derive(Clone, PartialEq)]
+pub struct CoreRange(pub(crate) RangeSetBlaze<usize>);
+
+impl Debug for CoreRange {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.0.to_string().as_str())
+    }
+}
+
+impl TryFrom<&[usize]> for CoreRange {
+    type Error = ParseError;
+
+    fn try_from(value: &[usize]) -> Result<Self, Self::Error> {
+        if value.is_empty() {
+            return Err(ParseError::EmptyRange);
+        }
+        Ok(CoreRange(RangeSetBlaze::from_iter(value)))
+    }
+}
+
+impl FromStr for CoreRange {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
```

Comment:
```suggestion
    /// Parse CoreRange from string like "1,2-30,31"
    fn from_str(s: &str) -> Result<Self, Self::Err> {
```

Btw, what happens on `1,1,1,2-30,2,2,2,3-10,64,1`? It will just work, without any duplication?

Reply:
yes